src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java

```diff
@@ -501,9 +501,73 @@ public synchronized boolean reserve(List<LockableResource> resources,
 		return true;
 	}
 
+	private void unreserveResources(@Nonnull List<LockableResource> resources) {
+		for (LockableResource l : resources) {
+			l.unReserve();
+		}
+		save();
+	}
 	public synchronized void unreserve(List<LockableResource> resources) {
+		// make sure there is a list of resources to unreserve
+		if (resources == null || (resources.size() == 0)) {
+			return;
+		}
+		List<String> resourceNamesToUnreserve = new ArrayList<>();
 		for (LockableResource r : resources) {
-			r.unReserve();
+			resourceNamesToUnreserve.add(r.getName());
+		}
+
+		// check if there are resources which can be unlocked (and shall not be unlocked)
+		List<LockableResource> requiredResourceForNextContext = null;
+		QueuedContextStruct nextContext = this.getNextQueuedContext(resourceNamesToUnreserve, false);
```

Comment:
So here's a thought: what if we do this a bit at a time? 

Basically we have the list of resources and a function that runs in a while loop and invokes this (or a slightly modified version of this) until it runs out of resources it can free or blocked items?

Should be basically just a big for loop that is a lot simpler and sidesteps most of the nastiness of this?